src/systemctl/systemctl.c

```diff
@@ -552,18 +553,45 @@ static int get_unit_list(
                         "org.freedesktop.systemd1",
                         "/org/freedesktop/systemd1",
                         "org.freedesktop.systemd1.Manager",
-                        "ListUnitsFiltered");
-
+                        "ListUnitsByStatesAndPatterns");
         if (r < 0)
                 return bus_log_create_error(r);
 
         r = sd_bus_message_append_strv(m, arg_states);
         if (r < 0)
                 return bus_log_create_error(r);
 
-        r = sd_bus_call(bus, m, 0, &error, &reply);
+        r = sd_bus_message_append_strv(m, patterns);
         if (r < 0)
-                return log_error_errno(r, "Failed to list units: %s", bus_error_message(&error, r));
+                return bus_log_create_error(r);
+
+        r = sd_bus_call(bus, m, 0, &error, &reply);
+        if (r < 0) {
+                // fallback to legacy ListUnitsFiltered method
+                fallback = true;
```

Comment:
Please check explicitly for the SD_BUS_ERROR_UNKNOWN_METHOD error here...

Nitpick: No // comments please, see CODING_STYLE